File: shinobicharts/Samples/Objective-C/HandlingSelection/HandlingSelection/PieChartDataSource.m
//
//  PieChartDataSource.m
//  HandlingSelection
//
//  Created by Colin Eberhardt on 16/07/2013.
//  Copyright (c) 2013 ShinobiControls. All rights reserved.
//

#import "PieChartDataSource.h"

@implementation PieChartDataSource
{
    NSDictionary* _sales;
}

- (id)initWithSales:(NSDictionary *)sales displayYear:(NSString *)year {
    if(self = [super init]) {
        _sales = sales;
        _displayYear = year;
    }
    return self;
}

- (NSDictionary*)dataForYear {
    NSDictionary* salesForYear = _sales[self.displayYear];
    return salesForYear;
}

#pragma mark - SChartDatasource methods

- (NSInteger)numberOfSeriesInSChart:(ShinobiChart *)chart {
    return 1;
}

-(SChartSeries *)sChart:(ShinobiChart *)chart seriesAtIndex:(NSInteger)index {
    SChartPieSeries *pieSeries = [[SChartPieSeries alloc] init];
    return pieSeries;
}

- (NSInteger)sChart:(ShinobiChart *)chart numberOfDataPointsForSeriesAtIndex:(NSInteger)seriesIndex {
   return [self dataForYear].count;
}

- (id<SChartData>)sChart:(ShinobiChart *)chart dataPointAtIndex:(NSInteger)dataIndex forSeriesAtIndex:(NSInteger)seriesIndex {
    SChartDataPoint *datapoint = [[SChartDataPoint alloc] init];
    
    NSDictionary* salesForYear = [self dataForYear];
    
    NSString* key = salesForYear.allKeys[dataIndex];
    datapoint.xValue = key;
    datapoint.yValue = salesForYear[key];
    return datapoint;
}

@end

File: shinobicharts/Samples/Swift/HandlingSelection/HandlingSelection/PieChartDatasource.swift
//
//  PieChartDatasource.swift
//  ShinobiControls
//
//  Copyright (c) 2014 Scott Logic. All rights reserved.
//

import UIKit

class PieChartDatasource: NSObject, SChartDatasource {
   
    // Currently selected year
    var displayYear: String
    
    let grocerySalesData: [String:[String:Double]]
    
    init(salesData: [String:[String:Double]], year: String) {
        self.grocerySalesData = salesData
        displayYear = year
    }
    
    // MARK:- Custom Functions
    
    // Get key value pairs for currently selected year
    func salesForDisplayYear() -> [String:Double]! {
        return grocerySalesData[displayYear]
    }
    
    // MARK:- SChartDatasource Functions
    
    // One series of sales data (current selected year in column series)
    func numberOfSeriesInSChart(chart: ShinobiChart!) -> Int {
        return 1
    }
    
    // Create pie series object
    func sChart(chart: ShinobiChart!, seriesAtIndex index: Int) -> SChartSeries! {
        return SChartPieSeries()
    }
    
    func sChart(chart: ShinobiChart!, numberOfDataPointsForSeriesAtIndex seriesIndex: Int) -> Int {
        return salesForDisplayYear().count
    }
    
    // Create data points representing segments of pie series
    func sChart(chart: ShinobiChart!, dataPointAtIndex dataIndex: Int, forSeriesAtIndex seriesIndex: Int) -> SChartData! {
        
        let salesForYear = salesForDisplayYear()
        
        let groceryType = salesForYear.keys.array[dataIndex]
        
        let dataPoint = SChartDataPoint()
        dataPoint.xValue = groceryType
        dataPoint.yValue = salesForYear[groceryType]
        
        return dataPoint
    }
    
}
